### PR TITLE
limit nodes fetching for 20 node per time to avoid jam, add likely down and loading status to nodes

### PR DESCRIPTION
**Pull request by waleedhammam**
#### Issue
https://github.com/threefoldtech/tfgridclient_proxy/issues/63
https://github.com/threefoldtech/tfgridclient_proxy/issues/66

#### Changelog
- limit nodes fetching for 20 node per time to avoid jam
- Introduce "likely down" status for nodes which can't communicated for 10 mins

### CODE REVIEW

File: internal/explorer/server.go [hunk 1]
```diff
@@ -175,17 +176,29 @@ func (a *App) getNode(w http.ResponseWriter, r *http.Request) {
...
+	response := NodeStatus{Status: "loading"}
+
 	nodeID := mux.Vars(r)["node_id"]
-	_, err := a.getNodeData(nodeID, false)
+	value, err := a.GetRedisKey(fmt.Sprintf("GRID3NODE:%s", nodeID))
 	if err != nil {
-		w.Header().Add("Content-Type", "application/json")
-		w.WriteHeader(http.StatusOK)
-		w.Write([]byte("{\"status\": \"down\"}"))
+		response.Status = "loading"
```
Comment (OmarElawady) [inline internal/explorer/server.go:184]: this doesn't separate errnotfound and other kinds of errors. when a node is not found it gets deleted from redis, so a non existent node would show state as loading?

File: internal/explorer/helpers.go [hunk 2]
```diff
@@ -287,21 +290,58 @@ func (a *App) getNodeData(nodeID string, force bool) (string, error) {
...
+				err = a.SetRedisKey(fmt.Sprintf("GRID3NODE:%s", nodeID), marshalledInfo, 10*60)
+				if err != nil {
+					log.Warn().Err(err).Msg("could not cache data in redis")
+				}
+
+				return "", ErrBadGateway
+			}
+
+			// down will be still down
+			if redisData.Status == "down" || redisData.Status == "" {
```
[the remark this answers is not in the document]
Comment (waleedhammam) [reply internal/explorer/helpers.go:315]: Likely down key will expire after 10 mins so will be wmpty
And loading is the status is empty
Here i check for empty includes both

File: internal/explorer/server.go (hunk 1 above)
Comment (waleedhammam) [reply internal/explorer/server.go:184]: No, that's until nodes loads, when the first loop complete it will be either up or down

**Comment by OmarElawady** [comment]
After discussion with @waleedhammam:
The old behavior was to return a "bad gateway" error when visiting `/node/<dead-node>` which made sense because it's an upstream server issue. In the pr, it returns empty data with "status: 'down'".

File: internal/explorer/helpers.go [hunk 3]
```diff
@@ -70,7 +72,7 @@ func (a *App) baseQuery(queryString string) (io.ReadCloser, error) {
 	if response.StatusCode != 200 {
 		var errResult interface{}
 		if err := json.NewDecoder(response.Body).Decode(errResult); err != nil {
-			return nil, fmt.Errorf("request failed: %w", errResult)
+			return nil, fmt.Errorf("request failed: %s", errResult)
```
Comment (muhamadazmy) [inline internal/explorer/helpers.go:75]: If the Decode method return an error, your errResult object will still be nil. what u probably meant to do is use `err` not `errResult` because that's what was returned by the Decode method.

In that case of course you should use the `%w` because it will wrap the error in the new error (which is a decoding error by the way, not a request failed)

```suggestion
			return nil, fmt.Errorf("failed to decode error from page: %w", err)
```

File: internal/explorer/helpers.go [hunk 4]
```diff
@@ -70,7 +72,7 @@ func (a *App) baseQuery(queryString string) (io.ReadCloser, error) {
...
 		}
 		return nil, fmt.Errorf("failed to query explorer network")
```
Comment (muhamadazmy) [inline internal/explorer/helpers.go:77]: Now here the decode of the return `Body` as error succeed, and now it's valid to return the `errResult` 
```suggestion
		return nil, fmt.Errorf("failed to query explorer network: %v", errResult)
```

File: internal/explorer/helpers.go [hunk 5]
```diff
@@ -287,21 +290,58 @@ func (a *App) getNodeData(nodeID string, force bool) (string, error) {
 			}
 			return "", ErrNodeNotFound
 		} else if err != nil {
-			err = a.DeleteRedisKey(fmt.Sprintf("GRID3NODE:%s", nodeID))
-			if err != nil {
-				log.Warn().Err(err).Msg("could not delete key in redis")
+
+			redisData := NodeInfo{}
+			json.Unmarshal([]byte(value), &redisData)
```
Comment (muhamadazmy) [inline internal/explorer/helpers.go:295]: Unhandled possible error. 

also I  reviewed the `fetchNodeData` method and you are leaking channels by creating them and never closing them
Pleas review this method again

File: internal/explorer/helpers.go [hunk 6]
```diff
@@ -287,21 +290,58 @@ func (a *App) getNodeData(nodeID string, force bool) (string, error) {
 			}
 			return "", ErrNodeNotFound
 		} else if err != nil {
-			err = a.DeleteRedisKey(fmt.Sprintf("GRID3NODE:%s", nodeID))
-			if err != nil {
-				log.Warn().Err(err).Msg("could not delete key in redis")
+
```
Comment (muhamadazmy) [inline internal/explorer/helpers.go:293]: Styling note, there is too much indentation in this method which makes it hard to review or read. You need to split it into smaller methods

File: internal/explorer/helpers.go [hunk 7]
```diff
@@ -287,21 +290,58 @@ func (a *App) getNodeData(nodeID string, force bool) (string, error) {
...
+			redisData := NodeInfo{}
+			json.Unmarshal([]byte(value), &redisData)
+
+			// up or loading make it likely down for 10 minutes
+			if redisData.Status == "up" {
+				// mark the node likely down if we can't reach this node in 10 mins it's down
+				nodeInfo.Status = "likely down"
+				marshalledInfo, err := json.Marshal(nodeInfo)
+				if err != nil {
+					log.Error().Err(err).Msg("could not marshal node info")
```
Comment (muhamadazmy) [inline internal/explorer/helpers.go:303]: I mentioned before that logging need to happen when u can't propagate the error anymore. When u need to continue, but as long as u return the error u shouldn't log. Because you then will end up logging the same error many times.

File: internal/explorer/helpers.go [hunk 8]
```diff
@@ -287,21 +290,58 @@ func (a *App) getNodeData(nodeID string, force bool) (string, error) {
...
-				log.Warn().Err(err).Msg("could not delete key in redis")
+
+			redisData := NodeInfo{}
+			json.Unmarshal([]byte(value), &redisData)
+
+			// up or loading make it likely down for 10 minutes
+			if redisData.Status == "up" {
+				// mark the node likely down if we can't reach this node in 10 mins it's down
+				nodeInfo.Status = "likely down"
+				marshalledInfo, err := json.Marshal(nodeInfo)
```
Comment (muhamadazmy) [inline internal/explorer/helpers.go:301]: Instead of always calling json.Marshal, and json.Unmarshal the nodeInfo type can has the methods to serialize, and deserialize which will make it easier (and cleaner) to change the encoding for example and make it more abstract

File: internal/explorer/helpers.go [hunk 9]
```diff
@@ -287,21 +290,58 @@ func (a *App) getNodeData(nodeID string, force bool) (string, error) {
...
+				}
+				err = a.SetRedisKey(fmt.Sprintf("GRID3NODE:%s", nodeID), marshalledInfo, 30*60)
+				if err != nil {
+					log.Warn().Err(err).Msg("could not cache data in redis")
+				}
+				return "", ErrBadGateway
 			}
+
+			log.Warn().Err(err).Msg(fmt.Sprintf("node %s is likely down", nodeID))
 			return "", ErrBadGateway
```
Comment (muhamadazmy) [inline internal/explorer/helpers.go:333]: Shouldn't this has the actual error that was returned by querying the node info. BadGateway is so generic and doesn't tell us much about the underlying error. You still can wrap the errors so it's a BadGateway but with details about what caused it.

File: internal/explorer/helpers.go (hunk 5 above)
Comment (muhamadazmy) [reply internal/explorer/helpers.go:295]: The fetchNodeData also does many calls to the node the DMI and hypervisor is never going to change. so u only need to get capacity and version. and I am actually thinking of creating one call that return both so u don't have to make multiple calls to the same node.

File: internal/explorer/helpers.go [hunk 10]
```diff
@@ -287,21 +290,58 @@ func (a *App) getNodeData(nodeID string, force bool) (string, error) {
...
 		}
 		// Save value in redis
-		// caching for 40 mins
+		// caching for 30 mins
+		nodeInfo.Status = "up"
 		marshalledInfo, err := json.Marshal(nodeInfo)
 		if err != nil {
 			log.Error().Err(err).Msg("could not marshal node info")
```
Comment (muhamadazmy) [inline internal/explorer/helpers.go:340]: Same comment on logging. and also the error you return is `internal server error` doesn't tell much about the cause. errors types and messages need to tell you the truth, otherwise debugging is gonna be so hard if u kept getting internal server error everywhere.

File: internal/explorer/server.go [hunk 11]
```diff
@@ -175,17 +176,31 @@ func (a *App) getNode(w http.ResponseWriter, r *http.Request) {
 func (a *App) getNodeStatus(w http.ResponseWriter, r *http.Request) {
 	enableCors(&w)
 
+	response := NodeStatus{Status: "loading"}
+
 	nodeID := mux.Vars(r)["node_id"]
-	_, err := a.getNodeData(nodeID, false)
+	value, err := a.GetRedisKey(fmt.Sprintf("GRID3NODE:%s", nodeID))
```
Comment (muhamadazmy) [inline internal/explorer/server.go:182]: I see this pattern a lot, wouldn't it be better if u have a method that `builds` the key for u so it's all in one place instead of copying the same string all over the place

```go
func getNodeKey(nodeID) string 
```

**Review by muhamadazmy — CHANGES_REQUESTED** [review]
General notes.

Avoid indentations, and if u have to, code in the indented block shouldn't be large because u lose track of what was the condition

For example
```go
if <condtion> {
  // so 
  // many 
  // logic
}

return error
```
can become 
```go
if ! <condtion> {
   return err
} 

// so 
// many 
// statements
```

In other words, shorter code paths can be indented but longer ones shouldn't

Second note is logging. excessive logging is more confusing, I would rather have descriptive errors over logs. You need to know when to log. If I log before every "return" then I will end up loggigng the same error multiple times (each time i bubble the error up the stack). Also the errors are renamed, or retyped is a bad habbit because your error returned to the caller then doesn't have enough information to tell what is going on.
So in conclusion:
- Log errors ONLY when there is no way to bubble the error up for example, in a loop where u need to process all nodes, but some might fail then u log the error and move on to next node on the list
- Always wrap your errors so they don't lose information for example
```go
if err := getNodeInfo(); err != nil {
   return fmt.Errorf("failed to get node info: %w", err)
}
```
This is imporant because, eventually the error will tell u exactly what happend, if getNodeInfo wraps the error, u would know that it happened because of connection error, or whatever. plus it will include information abou tthe node itself 

This also useful because u can unwrap the error and get underlying errors.

File: internal/explorer/helpers.go (hunk 5 above)
Comment (waleedhammam) [reply internal/explorer/helpers.go]: Will get them once and store it via lru

**Review by muhamadazmy — APPROVED** [review]

**Review by OmarElawady — APPROVED** [review]

File: internal/explorer/helpers.go [hunk 12]
```diff
@@ -216,13 +229,51 @@ func (a *App) getSystemVersion(ctx context.Context, nodeClient *client.NodeClien
...
+}
+
+func (a *App) getNodeDMI(ctx context.Context, nodeID string, nodeClient *client.NodeClient) (dmi.DMI, error) {
+	nodeKey := fmt.Sprintf("node_%s_dmi", nodeID)
+	if nodeDMI, found := a.lruCache.Get(nodeKey); found {
+		return nodeDMI.(dmi.DMI), nil
+	}
+
+	systemDMI := make(chan systemDMIReturnValue)
+	go a.getSystemDMI(ctx, nodeClient, systemDMI)
```
[the remark this answers is not in the document]
Comment (waleedhammam) [reply internal/explorer/helpers.go:260]: created an issue for it https://github.com/threefoldtech/tfgridclient_proxy/issues/69
